src/ireiat/data_pipeline/assets/rail_network/terminals.py

```diff
@@ -0,0 +1,242 @@
+
+    Returns:
+        ig.Graph: The updated impedance rail graph with terminal nodes and edges.
+    """
+    for row in terminals_df.itertuples(index=False):
+        node_idx = row.node_idx
+
+        # Skip rows with missing node_idx and log a warning
+        if pd.isnull(node_idx):
+            context.log.warning(f"Terminal {row.terminal_name} skipped: Missing node_idx.")
+            continue
+
+        terminal_operators: Set[str] = row.railroads
+
+        # Find vertices with the matching node index in the graph
+        vertices = impedance_rail_graph.vs.select(original_node_idx=node_idx)
+
+        # Track which terminal operators have been connected to a vertex
+        connected_railroads = set()
+        other_vertex = None  # Track if we find a vertex with 'Other' as the owner
+
+        # Check if we can connect the terminal to any existing vertices
+        for v in vertices:
+            vertex_owner = _extract_owner(v["original_idx"])
+
+            if vertex_owner in terminal_operators:
+                connected_railroads.add(vertex_owner)
+            elif vertex_owner == "Other" and other_vertex is None:
+                other_vertex = v
+
+        unconnected_railroads = set(terminal_operators) - connected_railroads
+
+        # If no valid connections (either to terminal operators or "Other" vertices), log a warning and skip this terminal
+        if not connected_railroads and not other_vertex:
+            context.log.warning(f"Terminal {row.terminal_name} skipped: No valid connections.")
+            continue
+
+        # Create two new vertices for the terminal (only if we know it will connect to the graph)
+        terminal_vertex_1 = impedance_rail_graph.add_vertex()
+        terminal_vertex_2 = impedance_rail_graph.add_vertex()
+
+        # Assign attributes to the new terminal vertices
+        terminal_vertex_1["terminal_name"] = row.terminal_name
+        terminal_vertex_1["vertex_type"] = "IM Terminal"
+        terminal_vertex_1["owners"] = terminal_operators
+        terminal_vertex_1["latitude"] = round(row.latitude, 6)
+        terminal_vertex_1["longitude"] = round(row.longitude, 6)
+
+        terminal_vertex_2["terminal_name"] = row.terminal_name
+        terminal_vertex_2["vertex_type"] = "IM Dummy Node"
+        terminal_vertex_2["owners"] = terminal_operators
+        terminal_vertex_2["latitude"] = round(row.latitude, 6)
+        terminal_vertex_2["longitude"] = round(row.longitude, 6)
+
+        # Add intermodal capacity edges between the two new vertices in both directions
+        im_capacity_edge_1 = impedance_rail_graph.add_edge(terminal_vertex_1, terminal_vertex_2)
+        im_capacity_edge_2 = impedance_rail_graph.add_edge(terminal_vertex_2, terminal_vertex_1)
+
+        # Assign attributes to the intermodal capacity edges
+        for im_capacity_edge in [im_capacity_edge_1, im_capacity_edge_2]:
+            im_capacity_edge["edge_type"] = "IM Capacity"
```

Comment:
We will probably need to refer to these edge types in the future. Would be good to put in an `Enum` class (or at least in `config.py` or something) to be able to refer to later...

Reply:
I've added an EdgeType Enum class inside util.rail_network_constants.py. Take a look and let me know if this is what you had in mind